hazelcast/src/main/java/com/hazelcast/map/impl/MapReplicationSupportingService.java

```diff
@@ -60,27 +63,34 @@ private void handleRemove(MapReplicationRemove replicationRemove) {
 
         try {
             int partitionId = nodeEngine.getPartitionService().getPartitionId(replicationRemove.getKey());
-            Future f = nodeEngine.getOperationService()
+            Future future = nodeEngine.getOperationService()
                     .invokeOnPartition(SERVICE_NAME, operation, partitionId);
-            f.get();
+            future.get();
         } catch (Throwable t) {
-            throw ExceptionUtil.rethrow(t);
+            throw rethrow(t);
         }
     }
 
     private void handleUpdate(MapReplicationUpdate replicationUpdate) {
-        EntryView<Data, Data> entryView = replicationUpdate.getEntryView();
-        MapMergePolicy mergePolicy = replicationUpdate.getMergePolicy();
+        Object mergePolicy = replicationUpdate.getMergePolicy();
         String mapName = replicationUpdate.getMapName();
         MapOperationProvider operationProvider = mapServiceContext.getMapOperationProvider(mapName);
-        MapOperation operation = operationProvider.createMergeOperation(mapName, entryView, mergePolicy, true);
+
+        MapOperation operation;
+        if (mergePolicy instanceof SplitBrainMergePolicy) {
```

Comment:
I think that's correct, but I will give it another round of thoughts later.

I guess the second case could be prevented with a cluster version check. I first thought those would be useless, since we have no idea about a target cluster. But in the RU scenario it might make sense. So if we have a new merge policy and the cluster version is below 3.10, we won't send an operation. Basically how we do it everywhere else as well.

Reply:
Unfortunately we don't have information on target cluster version so we can't avoid the first scenario.

Maybe you don't need to bother about the RU case either. It would cover the case when the ACTIVE cluster is lucky enough send the WAN event with the new merge policy to a 3.10 member which then detects that it is running in a 3.9 cluster. In other cases - when the update is sent to older members, the merge policy deserialisation will throw an exception.
And since the WAN updates are sent to all of the configured endpoints on the PASSIVE cluster, load balanced by event key partition ID, chances are pretty big one WAN update will be sent to a 3.9 member anyway.